### ISSUE TITLE
Static stats block with real time updating

### ISSUE THREAD

**Issue by doonze** [ncw/rclone#1210]
@ncw Nick, new issue per your request. A static stat block, like the current one, except it's persistent and only the information changes. I'll add my details later when I'm not on mobile, but just wanted to get the issue made for you.

So here's the low-down on the idea from the other issue:

The current stats block....

Transferred: 42.317 MBytes (119.671 kBytes/s)
Errors: 4
Checks: 28252
Transferred: 296
Elapsed time: 6m2s
Transferring:

var/lib/dpkg/info/libssh-4:armhf.list: 100% done, 522 Bytes/s, ETA: 0s
var/lib/dpkg/info/libssh-4:armhf.postrm: 100% done, 132 Bytes/s, ETA: 0s
var/lib/dpkg/info/libssh-4:armhf.md5sums: 100% done, 425 Bytes/s, ETA: 0s
var/lib/dpkg/info/libssh-4:armhf.postinst: 0% done, 0 Bytes/s, ETA: -
Last Error: List last error here.

I say print out this format as a static block if someone uses -stats, then update the stats in place (like --info=progress2 in rync). Adding the last error that was received at the bottom. Write off all errors to a file, then list them once the transfer is done.

Then once the transfer is done, the last block of stats and the errors are written off to output for peoples logs that use them.

This way you can get an update on what's going on at a glance, but you don't have a scrolling wall of text both onscreen and in your logs.

**Comment by j-vizcaino** [comment]
When implementing this, please consider that there should be a "log-ouput-friendly mode" for people who uses `rclone` in scripts and periodic tasks.
Thanks

**Comment by peterberbec** [comment]
100% behind this. I have a few precompiled, statically linked copies of the newest rsync running around to use on NAS boxes that have older versions w/o --info=progress2. 

My personal backup scripts have a large amount of code where I grep a tail of rclone's piped output. I arrive with one line I overwrite using bash cursor movement commands:

```
2017/04/17 00:02:07 - /encrypted 616002 files, 10808.3G
2017/04/17 00:02:28 - /encrypted 2058 files@5880/min, 44.9k@2.2kB/sec
```

**Comment by arekkas** [comment]
> My personal backup scripts have a large amount of code where I grep a tail of rclone's piped output. I arrive with one line I overwrite using bash cursor movement commands:

While this is not being addressed by rclone itself - can you provide a bit of info how you managed to pull this off? :)

**Comment by peterberbec** [comment]
Let me find it... It's kinda big and clunky

**Comment by peterberbec** [comment]
OK As requested. This is part of a bigger bash script. I call the upload function and pass it the directory I want to upload.

I'm going to start commenting the code now, but feel free to ask questions in the repo.

[rclonewatch](https://github.com/peterberbec/rclonewatch/blob/master/rclonewatch.sh)

**Comment by arekkas** [comment]
Awesome, thank you for sharing that!

**Comment by ncw** [comment]
I did make a start on this in #2347, from that thread

OK here is attempt number 2 within rclone which I think is getting there...  It would be activated with the --progress flag (like rsync).

[![asciicast](https://asciinema.org/a/189525.png)](https://asciinema.org/a/189525)

**Comment by arekkas** [comment]
That looks awesome already!

**Comment by peterberbec** [comment]
@ncw Two questions:

1. Can the progress output go to stdout and the other stuff go to stderr? The progress section keeps getting pushed down by the `$data: info : $filename copied : (new)` lines. Being able to pipe that somewhere else would be great. Ideally, the only thing that would output on stdout would be the progress stuff.

2. I love the work you've put into this, and I love all the details, but I love rsync --info=progress2's one-line approach. It aids in scripting if you're sure, combined with 1 above, there will only be one line of output.

Keep up the good work, and thanks for all you do. I, for one, welcome our new rclone overlords.

**Comment by ncw** [comment]
@peterberbec wrote:

> Can the progress output go to stdout and the other stuff go to stderr? The progress section keeps getting pushed down by the $data: info : $filename copied : (new) lines. Being able to pipe that somewhere else would be great. Ideally, the only thing that would output on stdout would be the progress stuff.

The idea is that  the log entries push the stats info to the bottom of the screen but it stays there as a non-moving block.  You can use `--log-file` to send the log elsewhere (that may not work right now though!)

> I love the work you've put into this, and I love all the details, but I love rsync --info=progress2's one-line approach. It aids in scripting if you're sure, combined with 1 above, there will only be one line of output.

For reference this is what `rsync --info=progress2` looks like

        482,712,279 100%   97.36MB/s    0:00:04 (xfr#1128, to-chk=0/1176)

It is showing total bytes transferred, speed, ETA, files to transfer and files to check.

rclone doesn't know some of those things but they would be nice to show certainly.

**Comment by peterberbec** [comment]
OK great.

Regarding rsync. It's not an exact copy of rsync I was hoping for, it was a single line of output, with whatever you were displaying

**Comment by ncw** [comment]
OK I've been working on getting the extra stats in.  What I'm imagining is a full stats block would look like this (this code is from here https://beta.rclone.org/branch/v1.42-055-g2d0c58a0-fix-379-stats/ if you want to give it a go).

```
Transferred:      20.111M / 335.136 MBytes, 6%, 10.017 MBytes/s, ETA 31s (xfr#74/784)
Errors:                 0
Checks:               356 / 356, 100%
Transferred:           74 / 784, 9%
Elapsed time:          2s
Transferring:
 *               puppies/dsc-20011020-161637.jpg: 87% /291.271k, 0/s, -
 *               puppies/dsc-20011020-161702.jpg: 89% /285.577k, 0/s, -
 *               puppies/dsc-20011020-161718.jpg: 33% /284.405k, 0/s, -
 *               puppies/dsc-20011020-161727.jpg: 11% /289.813k, 0/s, -
```

And a oneline stats block would be just (maybe with the `--stats-oneline` flag or something like that)

```
Transferred:      20.111M / 335.136 MBytes, 6%, 10.017 MBytes/s, ETA 31s (xfr#74/784)
```

Which has all the info rsync has in.

If you apply the `--progress` flag you'll get a static block which updates continously.

**Comment by peterberbec** [comment]
Nick, you're my hero! You've just allowed me to get rid of over a hundred lines of BS bash code.

I bought your wife some roses!

**Comment by peterberbec** [comment]
In the linked version of rclone, i get `Fatal error: unknown flag: --progress` and `Fatal error: unknown flag: --stats-oneline`. Is the code there but the option not enabled yet?

**Comment by ncw** [comment]
@peterberbec thank you!  I need to merge the progress branch with that one, and I haven't done the --stats-online flag yet (though it will be easy) - suggestions for a better name!

Expect all this to come together when I get back from my trip in a week!

**Comment by peterberbec** [comment]
`--rsync-stats`
`--progress2`
`--sls` # single line stats
`--pb-stats` lol

**Comment by ncw** [comment]
OK here is the next iteration. This has whole operation stats and supports --stats-one-line along with --progress.  I tried to make `--progress --one-line` look like rsync's `progress2` equivalent.

https://beta.rclone.org/branch/v1.42-126-g0475830a-fix-2347-progress/ (uploaded in 15-30 mins)

I haven't implemented @ewwink suggestion about switching the order of the transferring files and the transfer stats yet.

**Comment by ncw** [comment]
That one didn't build, try this one...

https://beta.rclone.org/branch/v1.42-128-gebf185ec-fix-2347-progress/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
3rd time lucky - this did build :-)

https://beta.rclone.org/branch/v1.42-128-g816d3ceb-fix-2347-progress/

**Comment by ncw** [comment]
I have merged that in

https://beta.rclone.org/v1.42-131-g1cccfa73/ (uploaded in 15-30 mins)

which will be in [the latest beta](https://beta.rclone.org) and v1.34

I expect that it will get  tweaked as time goes by!

**Comment by aeneasr** [comment]
Awesome! :)
